src/back/glsl.rs

```diff
@@ -38,16 +39,93 @@ impl fmt::Display for Error {
     }
 }
 
-pub fn write(module: &Module, out: &mut impl Write) -> Result<(), Error> {
-    writeln!(out, "#version 450 core")?;
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum Version {
+    Desktop(u16),
+    Embedded(u16),
+}
+
+impl fmt::Display for Version {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Version::Desktop(v) => write!(f, "{} core", v),
+            Version::Embedded(v) => write!(f, "{} es", v),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Options {
+    pub version: Version,
+    pub entry_point: (String, ShaderStage),
+}
+
+const SUPPORTED_CORE_VERSIONS: &[u16] = &[450, 460];
+const SUPPORTED_ES_VERSIONS: &[u16] = &[300, 310];
+
+bitflags::bitflags! {
+    struct SupportedFeatures: u32 {
+        const BUFFER_STORAGE = 1;
+        const SHARED_STORAGE = 1 << 1;
+        const SEPARATE_IMAGE_SAMPLER = 1 << 2;
+        const DOUBLE_TYPE = 1 << 3;
+        const NON_FLOAT_MATRICES = 1 << 4;
+        const MULTISAMPLED_TEXTURES = 1 << 5;
+        const MULTISAMPLED_TEXTURE_ARRAYS = 1 << 6;
+        const NON_2D_TEXTURE_ARRAYS = 1 << 7;
+    }
+}
+
+pub fn write(module: &Module, out: &mut impl Write, options: Options) -> Result<(), Error> {
+    let (version, es) = match options.version {
+        Version::Desktop(v) => {
+            if !SUPPORTED_CORE_VERSIONS.contains(&v) {
+                return Err(Error::Custom(format!(
+                    "Version not supported {}",
+                    options.version
+                )));
+            }
+
+            (v, false)
+        }
+        Version::Embedded(v) => {
+            if !SUPPORTED_ES_VERSIONS.contains(&v) {
+                return Err(Error::Custom(format!(
+                    "Version not supported {}",
+                    options.version
+                )));
+            }
+
+            (v, true)
+        }
+    };
+
+    writeln!(
+        out,
+        "#version {} {}",
+        version,
+        if es { "es" } else { "core" }
+    )?;
+
+    if es {
+        writeln!(out, "precision highp float;")?;
+    }
 
     let mut counter = 0;
     let mut names = FastHashMap::default();
 
     let mut namer = |name: Option<&String>| {
         if let Some(name) = name {
-            names.insert(name.clone(), ());
-            name.clone()
+            if name.starts_with("gl_") || name == "main" || names.get(name).is_some() {
```

Comment:
would it be simpler to have a `Map<&str, usize>`, which stores the last index used for each name? This way, we could use the borrowed strings for keys (instead of `String`) and avoid iteration altogether